tgui/packages/tgui/interfaces/Manufacturer/blueprintButton.tsx

```diff
@@ -0,0 +1,164 @@
+  materials_stored:Resource[]) =>
+{
+  let satisfaction:boolean[] = [];
+  let availableMaterials:Record<string, number> = {};
+  for (let i in pattern_requirements) {
+    let required_pattern = pattern_requirements[i];
+    let compatible_material:Resource;
+    // Try to find compatible resource
+    for (let resource of materials_stored) {
+      if (Object.keys(availableMaterials).find((value:string) => (value === resource.id)) === undefined) {
+        availableMaterials[resource.id] = resource.amount;
+      }
+      if (availableMaterials[resource.id] < amount_requirements[i]/10) {
+        continue;
+      }
+      if (resource.satisfies.find((satisfies_pattern:string) => (
+        (required_pattern === "ALL" || required_pattern === satisfies_pattern)
+      )) === undefined) {
+        continue;
+      }
+      compatible_material = resource;
+      break;
+    }
+    if (compatible_material === undefined) {
+      satisfaction.push(false);
+      continue;
+    }
+    satisfaction.push(true);
+    availableMaterials[compatible_material.id] -= amount_requirements[i];
+  }
+  return satisfaction;
+};
+
+export const BlueprintButton = (props, context) => {
+
+  const { blueprintData, materialData, manufacturerSpeed } = props;
+  const { act } = useBackend(context);
+  const blueprintSatisfaction:boolean[] = GetProductionSatisfaction(
+    blueprintData.item_paths,
+    blueprintData.item_amounts,
+    materialData
+  );
+  const isProduceable = blueprintSatisfaction.find((value:boolean) => !(value)) === undefined;
+  // Don't include this flavor if we only output one item, because if so, then we know what we're making
+  let outputs = (blueprintData.item_outputs.length < 2
+    && blueprintData.create < 2
+    && !blueprintData.apply_material) ? null : (
+      <>
+        <br />
+        Outputs: <br />
+        {blueprintData.item_names.map((value:string, index:number) => (
+          <b key={index}>
+            {blueprintData.create}x {value}<br />
+          </b>
+        ))}
+      </>
+    );
+  let content_requirements = (
+    <Section
+      title="Requirements"
+      buttons={<Button icon="hourglass" backgroundColor="rgba(0,0,0,0)">{getBlueprintTime(blueprintData.time, manufacturerSpeed)}s</Button>}
+    >
+      <LabeledList
+
+      >
+        {blueprintData.material_names.map((value:string, index:number) => (
+          <LabeledList.Item
+            key={index}
+            labelColor={(blueprintSatisfaction[index]) ? "label" : "red"}
+            label={value}
+            textAlign="right"
+          >
+            {blueprintData.item_amounts[index]/10}
+          </LabeledList.Item>
+        ))}
+      </LabeledList>
+      {outputs}
+    </Section>
+  );
+
+  // /datum/manufacture contains no description of its 'contents', so the first item works
+  let content_info = blueprintData.item_descriptions[0];
+  return (
+    <Stack inline
+      width={18.75}
+      height={5.5}
+      m={0.5}
```

Comment:
Top level components of a return providing margin is odd to me, as it's knowledge of how this component sits within its container that isn't actually part of the component. Consider letting the consumer pass in layout props instead.